explainer-datamodel.md

```diff
@@ -0,0 +1,413 @@
+
+### Primary Objects
+
+The data model for formatted text is quite simple and involves only two objects: `FormattedText`
+is a container that orders a list of `FormattedTextRun` objects that contain the actual text and
+formatting information. (The `FormattedText` container can also have formatting applied that may
+inherit down to the `FormattedTextRun`s.)
+
+To maximize authoring convenience, we are proposing using the 
+[ObserverableArray pattern](https://heycam.github.io/webidl/#idl-observable-array) to allow authors
+to manage the contained array exactly like a JavaScript Array.
+
+<img src="explainerresources/formatted-text-om.png" alt="FormattedText object holds an array of FormattedTextRun objects through a property called 'textruns'." align="center"/>
+
+## Creating FormattedText and FormattedTextRun objects
+
+`FormattedText` and `FormattedTextRun` objects are created using the 'new' pattern:
+
+```js
+let text = new FormattedText();
+let textrun = new FormattedTextRun();
+```
+
+Attach `FormattedTextRun` objects to their `FormattedText` object's `textruns` property (an Array):
+
+```js
+text.textruns.push( textrun );
+```
+
+Add text to the `FormattedTextRun` object:
+
+```js
+textrun.text = "hello";
+```
+
+To simplify creation of a `FormattedText` object and its textruns, the constructors also supports
+various overloads:
+
+```js
+let text = new FormattedText("hello", " world");
+// equivalent to:
+let text = new FormattedText();
+text.textruns.push( new FormattedTextRun( {text: "hello"} ) );
+text.textruns.push( new FormattedTextRun( " world" ) );
+// make a copy of all the text runs from another FormattedText object to be included in a new one:
+let text2 = new FormattedText( text.textruns );
+// copy everything from another FormattedText object and its text runs
+let text3 = new FormattedText( text2 );
+```
+
+## Common Styles
+
+`FormattedTextRun` objects have two styling shortcuts that can be set at construction time:
+* **font** - Any string accepted by the CSS text of the *font* property
+* **color** - Any string accepted by the CSS text of the *color* property
+
+```js
+let textrun2 = new FormattedTextRun( { text: "hello, I'm green with envy", font: "20pt Courier", color: "#00FF00" } );
+```
+These values can also be read and changed after the object is created:
+
+```js
+textrun2.font = "10pt Courier";
+textrun2.color = "#00CC00";
+```
+
+## CSS Styling of FormattedText and text runs
+A wide range of inline layout-related CSS is supported on both the `FormattedText` and
+`FormattedTextRun` objects.
+
+To set these use the `styleMap` property (just like in HTML):
+
+```js
+textrun2.styleMap.set("text-decoration", "underline");
+```
+
+The `styleMap` property is available on both the `FormattedText` and `FormattedTextRun` objects.
+
+It is also possible to set multiple styles at once similar to how HTML parses the `style` attribute:
+
+```js
+textrun2.setStyle( "color: yellow; font: 15pt Verdana" );
```

Comment:
This is not supported yet - can we do it as a later refinement, should we mark things that are not supported yet on the explainer ?

Reply:
See https://github.com/WICG/canvas-formatted-text/commit/742d9b4c4c1f454ffe47383aafb803a169bdae98